### ISSUE TITLE
Multiple copies of everything, and comments are out of date? What file do I actually /want/ (and how do you use it)?

### ISSUE THREAD

**Issue by fake-name** [devernay/cminpack#14]
I'm trying to port some python code that uses `hybrd` to C++, and finding it very hard to follow what's going on in this package:

There seems to be two versions of every file - `hybrd_.c` and `hybrd.c`, for example. I *think* that `hybrd_.c` is the one that's been significantly updated?  

http://devernay.free.fr/hacks/cminpack/ says this is a significant rework, but the [github readme](https://github.com/devernay/cminpack/blob/master/README.md) says:

> This is a C version of the minpack minimization package. It has been derived from the fortran code using f2c and some limited manual editing. 

As far as I can tell, http://devernay.free.fr/hacks/cminpack/ links to this github project, so is the github readme out of date?

Additionally, it seems like the comments in the various files have never been updated. I'm trying to figure out how to use the void\* facility:

> The function to be minimized and all the Minpack functions now take an extra "void*" argument, which can be used to pass any pointer-to-struct or pointer-to-class, and you can put all you extra parameters and data in that struct. Just cast this pointer to the appropriate pointer type in your function, and there they are! There is no need for global variables anymore. Be careful if you access the same object from different threads, though (a solution is to protect this extra data with a mutex).

But I'm getting lost somewhere in the `#define` macro maze, and the comments in `https://github.com/devernay/cminpack/blob/master/hybrd1.c` seem to be the ones from the original fortran. I also can't seem to find any documentation of the changes, or the new function signatures after all the precompiler munging.

------

It seems like it'd be optimal to take the original versions, which are kept for historical reasons (?), and at least put them in a subdirectory so they don't clutter up everything anymore.

There's a *lot* of precompiler magic. For example, the first parameter in `hybrd` is `__cminpack_decl_fcn_nn__`, but I have no idea where it's coming from. `cminpack.h` then `#defines` it to just `cminpack_func_nn fcn_nn`, neither component of which is defined anywhere, and I have no idea how this file ever compiles. I can only assume something *else* string-munges to `cminpack_func_nn` or `fcn_nn` (holy crap that's confusing).
  

For that matter, just naively reading the headers is confusing:

```
__cminpack_attr__
int CMINPACK_EXPORT __cminpack_func__(hybrd1)( __cminpack_decl_fcn_nn__
        void *p, int n, __cminpack_real__ *x, __cminpack_real__ *fvec, __cminpack_real__ tol,
        __cminpack_real__ *wa, int lwa );
```

Without preprocessing that, does it take the first parameter of ` __cminpack_decl_fcn_nn__ void *p`, where `__cminpack_decl_fcn_nn__` is some sort of accessibility specifier, or (as I infer from the documentation) is `__cminpack_decl_fcn_nn__` somehow a separate argument despite the lack of a comma? How does that work?

I don't know if I have a coherent question here aside from ".....wat", but in general, I think the documentation needs a little cleanup. I've used doxygen for generating "what it looks like in use" docs before, with great success. 

I'm gonna try to convince the precompiler to tell me what the functions I'm calling actually look like.

**Comment by fake-name** [comment]
Feeding it through the preprocessor:
```

int hybrd1( cminpack_func_nn fcn_nn,
        void *p, int n, double *x, double *fvec, double tol,
        double *wa, int lwa );
```

1. Ohhhhh.
2. WHY IS THE PREPROCESSOR ADDING COMMAS!!!!1111one!!!

**Comment by devernay** [comment]
Please see the examples. In your case `thybrdc.c`and `thybrd1c.c`. This is 
There are two versions of each function:
- the function with the FORTRAN API (with the trailing underscore), which are not here for historical reasons, but because some code might be using it (remember that this software is from 1980)
- the function with the C API (without the trailing underscore) which does exactly the same thing, but can also be used in a multi-threading context.
You don't *have* to use the extra void* parameter: it's here to pass extra data and avoid using global variables. Just pass NULL if you prefer global variables.
What you call "precompiler magic" is here to help with the different precisions (float, double half).  You don't have to care about this. All you have to do is `#define __cminpack_double__` before including `cminpack.h`, and write your code as in the examples, replacing `real` with `double` everywhere if you prefer.

**Comment by fake-name** [comment]
> the function with the FORTRAN API (with the trailing underscore), which are not here for historical reasons, but because some code might be using it (remember that this software is from 1980)

OK, but why aren't they in a separate directory ("current" / "legacy", or similar). I figured it out, but the point is that it's not documented *anywhere*.

> You don't have to use the extra void* parameter: it's here to pass extra data and avoid using global variables.

I specifically *do* want to use it, because it's exactly what I need (passing data down to the subroutine). My point was more the way all the preprocessor stuff is happening, you can't even figure out what the relevant function signatures look like until *after* the preprocessor has run, because the preprocessor is inserting commas into the function definitions. 

> All you have to do is `#define __cminpack_double__` before including `cminpack.h`, and write your code as in the examples, replacing real with double everywhere if you prefer.

Where is that documented? The example I'm working off ([hybdrv.c](https://github.com/devernay/cminpack/blob/master/examples/hybdrv.c) does `#define real __cminpack_real__` *after* `#include "cminpack.h"`.

**Comment by devernay** [comment]
See "Using CMinpack" on the project's page: http://devernay.free.fr/hacks/cminpack/

And also at the start of the header:
https://github.com/devernay/cminpack/blob/master/cminpack.h#L9

You don't even have to define anything if you just want double precision: `__cminpack_real__` is defined to double by default.

Feel free to write more documentation.
